### PR TITLE
feat(js): decodeIdToken

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

- Add utility function `decodeIdToken`
- Add basic type `IdTokenClaims`

Related #126 

## Linear Issue Reference

LOG-1355
LOG-1357

## Testing

- decode valid JWT → return claims
- decode invalid JWT → throw Error
- decode valid JWT without some claim (e.g. `iss`) → throw StructError

---
@logto-io/eng

### CODE REVIEW

File: packages/js/src/utils/id-token.ts
Comment (xiaoyijun) [inline packages/js/src/utils/id-token.ts:19]: nit:
```js
const [ _, encodedPayload ] = token.split('.');
```

File: packages/js/src/utils/id-token.ts
Comment (xiaoyijun) [inline packages/js/src/utils/id-token.ts:18]: Can we use the jose lib to decode ID token like @gao-sun mentioned in this [PR](https://github.com/logto-io/kotlin/pull/74#discussion_r787345337)?

Comment (gao-sun) [reply packages/js/src/utils/id-token.ts:18]: @IceHe tested, unfortunately jose cannot do the job.

File: packages/js/src/utils/id-token.test.ts
Comment (gao-sun) [inline packages/js/src/utils/id-token.test.ts:25]: as i mentioned we also need to test if the provider returns more claims than we defined

File: packages/js/src/utils/id-token.ts
Comment (IceHe) [reply packages/js/src/utils/id-token.ts:19]: eslint doesn't recommend me to use code above.

image

File: packages/js/src/utils/id-token.ts
Comment (gao-sun) [inline packages/js/src/utils/id-token.ts:31]: why do we still have `JSON.parse` here?

File: packages/js/src/utils/id-token.ts
Comment (gao-sun) [inline packages/js/src/utils/id-token.ts:38]: no need for try - catch 😂

**Review by gao-sun — CHANGES_REQUESTED** [review]

File: packages/js/src/utils/id-token.ts
Comment (IceHe) [reply packages/js/src/utils/id-token.ts:19]: https://files.slack.com/files-pri/T0243NVUC9E-F02UZQYRQJV/image.png

Comment (IceHe) [reply packages/js/src/utils/id-token.ts:19]: image

File: packages/js/src/utils/id-token.ts
Comment (IceHe) [reply packages/js/src/utils/id-token.ts:38]: Removed, PTAL.

File: packages/js/src/utils/id-token.ts
Comment (IceHe) [reply packages/js/src/utils/id-token.ts:31]: rewrote like this, PTAL:
```ts
  const json = UrlSafeBase64.decode(encodedPayload);
  return s.create(json, IdTokenClaimsSchema);
}
```

File: packages/js/src/utils/id-token.ts
Comment (simeng-li) [inline packages/js/src/utils/id-token.ts:22]: Could refer to the Error definitions in Kotlin or Swift SDKs. Could define multiple type of errors for specific exceptions case. e.g. throw new JWTError('Invalid JWT format'). or throw new JWTError('Invalid token claims')

P.S Do we have a test case for this?

File: packages/js/src/utils/id-token.ts
Comment (simeng-li) [reply packages/js/src/utils/id-token.ts:31]: I think the `json` value from `UrlSafeBase64.decode` is a string. Not sure we could bypass the JSON.parse.

File: packages/js/src/utils/id-token.ts
Comment (simeng-li) [reply packages/js/src/utils/id-token.ts:38]: This is based on the previous comments. The try..catch might be valid if we have to keep JSON.parse and throw a controllable error type.

File: packages/js/src/utils/id-token.ts
Comment (IceHe) [reply packages/js/src/utils/id-token.ts]: Still use `JSON.parse` according to JOSE implementation.

File: packages/js/src/utils/id-token.test.ts
Comment (IceHe) [reply packages/js/src/utils/id-token.test.ts:25]: Added a new test case 'decoding valid JWT with non-predefined claims should return all claims', PTAL.

File: packages/js/src/utils/id-token.ts
Comment (IceHe) [reply packages/js/src/utils/id-token.ts]: The variable named `json` can also present a string value containing json content. e.g. demo code from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse

image

File: packages/js/src/utils/id-token.ts
Comment (IceHe) [reply packages/js/src/utils/id-token.ts]: @wangsijie Any idea?

Comment (IceHe) [reply packages/js/src/utils/id-token.ts]: Conclusion: No need for try-catch ( See [discussion](https://silverhand-io.slack.com/archives/C02A8G4HVAM/p1642588712357800) )

File: packages/js/src/utils/id-token.ts
Comment (IceHe) [reply packages/js/src/utils/id-token.ts:21]: Conclusion: Do not define custom Error for now, and do it when needed ( See [discussion](https://silverhand-io.slack.com/archives/C02A8G4HVAM/p1642588826359200)

File: packages/js/package.json
Comment (gao-sun) [inline packages/js/package.json:22]: this should be put in `devDependencies` for now, or users will install `jose` while they don't need

File: packages/js/src/utils/id-token.ts
Comment (gao-sun) [reply packages/js/src/utils/id-token.ts:21]: still need a custom error here. would you like to add a linear issue before merging?

**Review by gao-sun — APPROVED** [review]
overall lgtm, approve to unblock

**Comment by IceHe** [comment]
> @gao-sun : still need a custom error here. would you like to add a linear issue before merging?

Added LOG-1355

File: packages/js/package.json
Comment (IceHe) [reply packages/js/package.json:22]: My bad.

But it will be required by `verifyIdToken` in the next PR.
I prefer keep it for now.

File: packages/js/src/utils/id-token.ts
Comment (IceHe) [reply packages/js/src/utils/id-token.ts:21]: Added LOG-1355

**Pull request merged**
